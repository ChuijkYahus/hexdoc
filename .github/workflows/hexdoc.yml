name: "[Reusable] Build and publish a hexdoc project"

on:
  workflow_call:
    inputs:
      python-version:
        description: Python version to install
        type: string
        required: true
      checkout:
        description: Whether to check out the branch specified in `branch`
        type: boolean
        required: true
      branch:
        description: Branch to generate docs from
        type: string
        required: false
      publish-latest-and-root:
        description: Overwrite `/v/latest` (and `/`, if releasing)
        type: boolean
        required: true
      publish-release:
        description: Overwrite `/v/<current_version>`
        type: boolean
        required: true
      bump-version-segment:
        description: Version segment to bump with Hatch
        type: string
        required: false
      props-path:
        description: Path to your props file (hexdoc.toml or properties.toml)
        type: string
        required: false
      pip-extras:
        description: Pip extras for your package (eg. [dev])
        type: string
        required: false
    outputs:
      publish-release:
        description: Value of publish-release input for convenience.
        value: ${{ inputs.publish-release }}
      pages-url:
        description: Current GitHub Pages url for this repo.
        value: ${{ jobs.build.outputs.pages-url }}

permissions:
  contents: read

env:
  UPDATE_LATEST: ${{ inputs.publish-latest-and-root }}
  RELEASE: ${{ inputs.publish-release }}

jobs:
  build:
    runs-on: ubuntu-latest

    env:
      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

    permissions:
      contents: write
      pages: read

    outputs:
      pages-url:  ${{ steps.get-url.outputs.pages-url }}
      matrix:     ${{ steps.list-langs.outputs.matrix }}
      branch:     ${{ steps.parse-inputs.outputs.branch }}
      props-path: ${{ steps.parse-inputs.outputs.props-path }}

    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-python@v4
        with:
          python-version: ${{ inputs.python-version }}
          cache: pip

      - name: Get Pages url
        id: get-url
        run: |
          url=$(gh api "repos/$GITHUB_REPOSITORY/pages" --jq '.html_url')
          echo "pages-url=$url" >> "$GITHUB_OUTPUT"
          echo "GITHUB_PAGES_URL=$url" >> "$GITHUB_ENV"

      - name: Parse inputs
        id: parse-inputs
        run: |
          if [[ ${{ inputs.checkout }} == true ]]; then
            branch='${{ inputs.branch }}'
            props_path=_checkout/${{ inputs.props-path }}
            echo "HATCH_GRADLE_DIR=_checkout" >> "$GITHUB_ENV"
          else
            branch=none
            props_path=${{ inputs.props-path }}
          fi

          echo "HEXDOC_PROPS=$props_path" >> "$GITHUB_ENV"

          echo "branch=$branch"         >> "$GITHUB_OUTPUT"
          echo "props-path=$props_path" >> "$GITHUB_OUTPUT"

      - name: Checkout input branch
        if: steps.parse-inputs.outputs.branch != 'none'
        uses: actions/checkout@v3
        with:
          ref: ${{ steps.parse-inputs.outputs.branch }}
          path: _checkout

      - name: Install docgen from source
        run: pip install .${{ inputs.pip-extras }} hatch

      - name: List book languages
        id: list-langs
        run: |
          echo "matrix=$(hexdoc list-langs)" >> "$GITHUB_OUTPUT"
          if [[ $RUNNER_DEBUG ]]; then
            tree -I '__pycache__|Common|Fabric|Forge|venv'
          fi

      - name: Export hexdoc resources
        run: hexdoc export

      - name: Bump version
        if: inputs.bump-version-segment
        run: hatch version "${{ inputs.bump-version-segment }}"

      - name: Commit changes
        uses: stefanzweifel/git-auto-commit-action@v4
        with:
          commit_message: Build web book from ${{ github.ref }}

      - name: Build package
        run: hatch build

      - name: Upload package artifact
        uses: actions/upload-artifact@v3
        with:
          name: hexdoc-build
          path: dist

      - name: Copy build to Pages
        if: env.UPDATE_LATEST
        run: |
          mkdir -p _site/v/latest
          cp -r dist _site/v/latest/dist

      - name: Upload temporary Pages artifact
        if: env.UPDATE_LATEST
        uses: actions/upload-artifact@v3
        with:
          name: github-pages-tmp
          path: _site

      - name: Add job summary
        run: echo "Built version \`$(hatch version)\` from commit \`$(git rev-parse --short "$GITHUB_SHA")\`." >> $GITHUB_STEP_SUMMARY

  generate:
    runs-on: ubuntu-latest
    needs: build
    continue-on-error: true

    strategy:
      fail-fast: false
      matrix:
        lang: ${{ fromJson(needs.build.outputs.matrix) }}

    env:
      GITHUB_PAGES_URL: ${{ needs.build.outputs.pages-url }}
      HEXDOC_PROPS: ${{ needs.build.outputs.props-path }}
      HEXDOC: hexdoc render _site --lang ${{ matrix.lang }} --clean

    steps:
      - uses: actions/checkout@v3
      - uses: actions/checkout@v3
        with:
          repository: object-Object/hexdoc
          path: _checkout_object-Object_hexdoc
      - uses: ./_checkout_object-Object_hexdoc/.github/actions/install-artifact-wheel
        with:
          name: hexdoc-build
          python-version: ${{ inputs.python-version }}
          pip-extras: ${{ inputs.pip-extras }}

      - name: Checkout input branch
        if: needs.build.outputs.branch != 'none'
        uses: actions/checkout@v3
        with:
          ref: ${{ needs.build.outputs.branch }}
          path: _checkout

      - name: Generate web book
        id: gen-normal
        continue-on-error: true
        run: $HEXDOC

      - name: Generate web book with missing translations
        if: steps.gen-normal.outcome == 'failure'
        run: $HEXDOC --allow-missing

      - name: Upload temporary Pages artifact
        uses: actions/upload-artifact@v3
        with:
          name: github-pages-tmp
          path: _site

      - name: Fail if the first generate step failed
        if: steps.gen-normal.outcome == 'failure'
        run: |
          echo "::error::Missing some i18n keys."
          exit 1

  deploy-pages:
    runs-on: ubuntu-latest
    needs: [build, generate]

    permissions:
      contents: write

    steps:
      - uses: actions/checkout@v3
      - uses: actions/checkout@v3
        with:
          repository: object-Object/hexdoc
          path: _checkout_object-Object_hexdoc
      - uses: ./_checkout_object-Object_hexdoc/.github/actions/install-artifact-wheel
        with:
          name: hexdoc-build
          python-version: ${{ inputs.python-version }}
          pip-extras: ${{ inputs.pip-extras }}

      - name: Checkout current Pages
        uses: actions/checkout@v3
        continue-on-error: true
        with:
          ref: gh-pages
          path: _site/dst

      - name: Download temporary Pages artifact
        uses: actions/download-artifact@v3
        with:
          name: github-pages-tmp
          path: _site/src/docs

      - name: Add new docs to site
        run: hexdoc merge --src _site/src/docs --dst _site/dst/docs

      - name: Deploy to Pages
        uses: JamesIves/github-pages-deploy-action@v4
        with:
          folder: _site/dst/docs
          target-folder: docs
