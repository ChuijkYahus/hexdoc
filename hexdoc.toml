modid = "hexdoc"
book = { "!None"={} }
default_lang = "en_us"

_export_root = "src/hexdoc/_export"
resource_dirs = [  # top takes priority
    { path="{^_export_root}/resources", reexport=false },
]
export_dir = "{_export_root}/generated"

template = { "!None"={} }

[minecraft_assets]
# https://github.com/PrismarineJS/minecraft-assets/tree/83e2169afbbce40990d69fc53e5962e4a793d467/data/1.19.1
ref = "83e2169afbbce40990d69fc53e5962e4a793d467"
version = "1.19.1"
