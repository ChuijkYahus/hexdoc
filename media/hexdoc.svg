<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
   xmlns:dc="http://purl.org/dc/elements/1.1/"
   xmlns:cc="http://creativecommons.org/ns#"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   xmlns:svg="http://www.w3.org/2000/svg"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   sodipodi:docname="hexdoc.svg"
   inkscape:version="1.0 (4035a4fb49, 2020-05-01)"
   inkscape:export-ydpi="48"
   inkscape:export-xdpi="48"
   inkscape:export-filename="C:\Users\jack\Pictures\hexdoc.png"
   id="svg8"
   version="1.1"
   viewBox="0 0 135.46666 135.46667"
   height="512"
   width="512">
  <defs
     id="defs2" />
  <sodipodi:namedview
     inkscape:window-maximized="1"
     inkscape:window-y="518"
     inkscape:window-x="-8"
     inkscape:window-height="1417"
     inkscape:window-width="2560"
     inkscape:snap-object-midpoints="false"
     inkscape:guide-bbox="true"
     showguides="true"
     units="px"
     showgrid="false"
     inkscape:document-rotation="0"
     inkscape:current-layer="layer1"
     inkscape:document-units="px"
     inkscape:cy="430.7964"
     inkscape:cx="5.0637413"
     inkscape:zoom="0.7"
     inkscape:pageshadow="2"
     inkscape:pageopacity="0.0"
     borderopacity="1.0"
     bordercolor="#666666"
     pagecolor="#ffffff"
     id="base">
    <sodipodi:guide
       id="guide837"
       orientation="0,512"
       position="0,135.46666" />
    <sodipodi:guide
       id="guide839"
       orientation="512,0"
       position="135.46666,135.46666" />
    <sodipodi:guide
       id="guide841"
       orientation="0,-512"
       position="135.46666,0" />
    <sodipodi:guide
       id="guide843"
       orientation="-512,0"
       position="0,0" />
    <sodipodi:guide
       inkscape:color="rgb(0,0,255)"
       inkscape:locked="false"
       inkscape:label=""
       id="guide847"
       orientation="-1,0"
       position="67.733332,67.733332" />
  </sodipodi:namedview>
  <metadata
     id="metadata5">
    <rdf:RDF>
      <cc:Work
         rdf:about="">
        <dc:format>image/svg+xml</dc:format>
        <dc:type
           rdf:resource="http://purl.org/dc/dcmitype/StillImage" />
        <dc:title></dc:title>
      </cc:Work>
    </rdf:RDF>
  </metadata>
  <g
     id="layer1"
     inkscape:groupmode="layer"
     inkscape:label="Layer 1">
    <path
       d="M 116.56666,95.94048 67.73333,124.14763 18.900005,95.94048 V 39.52618 L 67.73333,11.31903 116.56666,39.526181 Z"
       style="fill:#332233;fill-opacity:1;stroke:#201a20;stroke-width:16.9333;stroke-linecap:square;stroke-linejoin:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1;paint-order:stroke fill markers"
       id="path835" />
    <path
       id="path850"
       d="M 42.278415,67.733333 H 93.188249"
       style="fill:none;stroke:#dddddd;stroke-width:8.46667;stroke-linecap:round;stroke-linejoin:miter;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1" />
    <path
       style="fill:none;stroke:#dddddd;stroke-width:8.46667;stroke-linecap:round;stroke-linejoin:miter;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1"
       d="M 42.278416,87.771403 H 93.188249"
       id="path852" />
    <path
       id="path854"
       d="M 42.278416,47.695261 H 93.188249"
       style="fill:none;stroke:#dddddd;stroke-width:8.46667;stroke-linecap:round;stroke-linejoin:miter;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1" />
  </g>
</svg>
