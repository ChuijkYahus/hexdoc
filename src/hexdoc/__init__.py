"""
A [Jinja](https://jinja.palletsprojects.com/en/3.1.x/)-based documentation generator for [Patchouli](https://github.com/VazkiiMods/Patchouli) books.
"""

# attempt to reduce the amount of breaking changes from moving hexdoc out of hexmod
try:
    import hexdoc_hexcasting as hexcasting  # type: ignore
except ImportError:
    pass
