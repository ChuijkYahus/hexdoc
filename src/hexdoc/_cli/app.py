import code
import json
import logging
import os
import shutil
import sys
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from textwrap import dedent
from typing import Annotated, Union

from typer import Option, Typer

from hexdoc.core import ModResourceLoader
from hexdoc.minecraft import I18n
from hexdoc.minecraft.assets import AnimatedTexture

from . import render_block
from .utils.args import (
    DEFAULT_MERGE_DST,
    DEFAULT_MERGE_SRC,
    PathArgument,
    PropsOption,
    ReleaseOption,
    UpdateLatestOption,
    VerbosityOption,
)
from .utils.load import load_book, load_book_and_loader, load_books, load_common_data
from .utils.logging import repl_readfunc
from .utils.render import create_jinja_env, render_book
from .utils.sitemap import (
    assert_version_exists,
    delete_root_book,
    delete_updated_books,
    dump_sitemap,
    load_sitemap,
)

app = Typer(
    pretty_exceptions_enable=False,
    context_settings={
        "help_option_names": ["--help", "-h"],
    },
)
app.add_typer(render_block.app)


@app.command()
def list_langs(
    *,
    props_file: PropsOption,
    verbosity: VerbosityOption = 0,
):
    """Get the available language codes as a JSON list."""
    props, pm, _ = load_common_data(props_file, verbosity)
    with ModResourceLoader.load_all(
        props, pm, texture_render_dir=None, export=False
    ) as loader:
        langs = sorted(I18n.list_all(loader))
        print(json.dumps(langs))


@app.command()
def repl(
    *,
    props_file: PropsOption,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Start a Python shell with some helpful extra locals added from hexdoc."""
    props, pm, _ = load_common_data(props_file, verbosity)

    _, book, i18n, loader, context = load_book_and_loader(
        props.book,
        props,
        pm,
        lang,
        allow_missing,
        export=False,
        texture_render_dir=Path("out/tmp/textures"),
    )

    repl_locals = dict(
        props=props,
        pm=pm,
        book=book,
        i18n=i18n,
        loader=loader,
        context=context,
    )

    code.interact(
        banner=dedent(
            f"""\
            [hexdoc repl] Python {sys.version}
            Locals: {', '.join(sorted(repl_locals.keys()))}"""
        ),
        readfunc=repl_readfunc(),
        local=repl_locals,
        exitmsg="",
    )


@app.command()
def export(
    *,
    props_file: PropsOption,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Run hexdoc, but skip rendering the web book - just export the book resources."""
    props, pm, _ = load_common_data(props_file, verbosity)
    load_book(
        props.book,
        props,
        pm,
        lang,
        allow_missing,
        export=True,
        texture_render_dir=None,
    )


@app.command()
def render(
    output_dir: PathArgument = DEFAULT_MERGE_SRC,
    *,
    props_file: PropsOption,
    update_latest: UpdateLatestOption = True,
    release: ReleaseOption = False,
    clean: bool = False,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Export resources and render the web book."""

    # load data
    props, pm, version = load_common_data(props_file, verbosity)
    if not props.book:
        raise ValueError("Expected a value for props.book, got None")
    if not props.template:
        raise ValueError("Expected a value for props.template, got None")

    books, all_metadata = load_books(
        props.book,
        props,
        pm,
        lang,
        allow_missing,
        texture_render_dir=output_dir / "textures",
    )

    logger = logging.getLogger(__name__)
    logger.info(f"update_latest={update_latest}, release={release}")

    # set up Jinja
    env = create_jinja_env(pm, props.template.include, props_file)

    template_names = (
        props.template.render
        if props.template.override_default_render
        else pm.default_rendered_templates(props.template.include)
    ) | props.template.extend_render

    templates = {
        Path(path): env.get_template(template_name)
        for path, template_name in template_names.items()
    }

    if not templates:
        raise RuntimeError(
            "No templates to render, check your props.template configuration "
            f"(in {props_file.as_posix()})"
        )

    if clean:
        shutil.rmtree(output_dir, ignore_errors=True)

    lang_names = {
        lang_: f"{i18n.localize('language.name')} ({i18n.localize('language.region')})"
        for lang_, (_, i18n, _) in books.items()
    }

    for should_render, version_, is_root in [
        (update_latest, "latest", False),
        (release, version, False),
        (update_latest and release, version, True),
    ]:
        if not should_render:
            continue
        for lang_, (book, i18n, context) in books.items():
            textures = context.png_textures
            animations = sorted(
                (
                    texture
                    for texture in textures.values()
                    if isinstance(texture, AnimatedTexture)
                ),
                key=lambda t: t.class_name,
            )

            render_book(
                props=props,
                pm=pm,
                lang=lang_,
                lang_names=lang_names,
                book=book,
                i18n=i18n,
                templates=templates,
                output_dir=output_dir,
                all_metadata=all_metadata,
                textures=textures,
                animations=animations,
                allow_missing=allow_missing,
                version=version_,
                is_root=is_root,
            )

    logger.info("Done.")


@app.command()
def merge(
    *,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    update_latest: UpdateLatestOption = True,
    release: ReleaseOption = False,
):
    # ensure at least the default language was built successfully
    if update_latest:
        assert_version_exists(root=src, version="latest")

    # TODO: figure out how to do this with pluggy (we don't have the props file here)
    # if is_release:
    #     assert_version_exists(src, GRADLE_VERSION)

    dst.mkdir(parents=True, exist_ok=True)

    # remove any stale data that we're about to replace
    delete_updated_books(src=src, dst=dst)
    if update_latest and release:
        delete_root_book(root=dst)

    # do the merge
    shutil.copytree(src=src, dst=dst, dirs_exist_ok=True)

    # rebuild the sitemap
    sitemap = load_sitemap(dst)
    dump_sitemap(dst, sitemap)


@app.command()
def serve(
    *,
    props_file: PropsOption,
    port: int = 8000,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    do_merge: Annotated[bool, Option("--merge/--no-merge")] = False,
    update_latest: bool = True,
    release: bool = True,  # you'd generally want --release for development
    clean: bool = False,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    book_root = dst if do_merge else src

    book_path = book_root.resolve().relative_to(Path.cwd())

    book_url = f"/{book_path.as_posix()}"

    os.environ |= {
        "DEBUG_GITHUBUSERCONTENT": "",
        "GITHUB_PAGES_URL": book_url,
    }

    print("Rendering...")
    render(
        props_file=props_file,
        output_dir=src,
        update_latest=update_latest,
        release=release,
        clean=clean,
        lang=lang,
        allow_missing=allow_missing,
        verbosity=verbosity,
    )

    if do_merge:
        print("Merging...")
        merge(
            src=src,
            dst=dst,
            update_latest=update_latest,
            release=release,
        )

    print(
        f"Serving web book at http://localhost:{port}{book_url} (press ctrl+c to exit)\n"
    )
    with HTTPServer(("", port), SimpleHTTPRequestHandler) as httpd:
        # ignore KeyboardInterrupt to stop Typer from printing "Aborted."
        # because it keeps printing after nodemon exits and breaking the output
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            pass


if __name__ == "__main__":
    app()
