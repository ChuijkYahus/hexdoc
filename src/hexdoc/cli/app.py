import code
import json
import logging
import os
import shutil
import sys
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from textwrap import dedent
from typing import Annotated, Union

from typer import Option, Typer

from hexdoc.core import ModResourceLoader
from hexdoc.minecraft import I18n, Tag
from hexdoc.minecraft.assets import (
    AnimatedTexture,
    HexdocAssetLoader,
    PNGTexture,
)
from hexdoc.utils.git import git_root

from . import render_block
from .utils.args import (
    DEFAULT_MERGE_DST,
    DEFAULT_MERGE_SRC,
    BranchOption,
    PathArgument,
    PropsOption,
    ReleaseOption,
    VerbosityOption,
)
from .utils.load import (
    load_book,
    load_common_data,
    render_textures_and_export_metadata,
)
from .utils.logging import repl_readfunc
from .utils.render import create_jinja_env, render_book
from .utils.sitemap import (
    assert_version_exists,
    delete_root_book,
    delete_updated_books,
    dump_sitemap,
    load_sitemap,
)

logger = logging.getLogger(__name__)

app = Typer(
    pretty_exceptions_enable=False,
    context_settings={
        "help_option_names": ["--help", "-h"],
    },
)
app.add_typer(render_block.app)


@app.command()
def list_langs(
    *,
    props_file: PropsOption,
    verbosity: VerbosityOption = 0,
):
    """Get the available language codes as a JSON list."""
    props, pm, _ = load_common_data(props_file, verbosity, "")
    with ModResourceLoader.load_all(
        props,
        pm,
        export=False,
    ) as loader:
        langs = sorted(I18n.list_all(loader))
        print(json.dumps(langs))


@app.command()
def repl(
    *,
    props_file: PropsOption,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Start a Python shell with some helpful extra locals added from hexdoc."""
    props, pm, _ = load_common_data(props_file, verbosity, "")

    loader = ModResourceLoader.load_all(
        props,
        pm,
        export=False,
    )

    _, book, i18n, context = load_book(
        props.book,
        props,
        pm,
        loader,
        lang,
        allow_missing,
    )

    repl_locals = dict(
        props=props,
        pm=pm,
        book=book,
        i18n=i18n,
        loader=loader,
        context=context,
    )

    code.interact(
        banner=dedent(
            f"""\
            [hexdoc repl] Python {sys.version}
            Locals: {', '.join(sorted(repl_locals.keys()))}"""
        ),
        readfunc=repl_readfunc(),
        local=repl_locals,
        exitmsg="",
    )


@app.command()
def export(
    output_dir: PathArgument = DEFAULT_MERGE_SRC,
    *,
    branch: BranchOption,
    release: ReleaseOption = False,
    props_file: PropsOption,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Run hexdoc, but skip rendering the web book - just export the book resources."""
    props, pm, plugin = load_common_data(props_file, verbosity, branch)

    loader = ModResourceLoader.clean_and_load_all(
        props,
        pm,
        export=True,
    )

    site_path = plugin.site_path(versioned=release)
    asset_loader = plugin.asset_loader(
        loader=loader,
        site_url=f"{loader.props.url}/{site_path.as_posix()}",  # TODO: urlencode
        asset_url=props.env.asset_url,
        render_dir=output_dir / site_path,
    )

    all_metadata = render_textures_and_export_metadata(loader, asset_loader)

    i18n = I18n.load_all(
        loader,
        allow_missing=allow_missing,
    )[props.default_lang]

    if props.book:
        load_book(
            book_id=props.book,
            pm=pm,
            loader=loader,
            i18n=i18n,
            all_metadata=all_metadata,
        )


@app.command()
def render(
    output_dir: PathArgument = DEFAULT_MERGE_SRC,
    *,
    branch: BranchOption,
    release: ReleaseOption = False,
    clean: bool = False,
    lang: Union[str, None] = None,
    props_file: PropsOption,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Export resources and render the web book."""

    # load data
    props, pm, plugin = load_common_data(props_file, verbosity, branch, book=True)

    if not props.book:
        raise ValueError("Expected a value for props.book, got None")
    if not props.template:
        raise ValueError("Expected a value for props.template, got None")

    site_path = plugin.site_book_path(
        lang=lang or props.default_lang,
        versioned=release,
    )

    books, all_metadata = load_books(
        props.book,
        props,
        pm,
        lang,
        allow_missing,
    )

    # set up Jinja
    env = create_jinja_env(pm, props.template.include, props_file)

    template_names = (
        props.template.render
        if props.template.override_default_render
        else pm.default_rendered_templates(props.template.include)
    ) | props.template.extend_render

    templates = {
        Path(path): env.get_template(template_name)
        for path, template_name in template_names.items()
    }

    if not templates:
        raise RuntimeError(
            "No templates to render, check your props.template configuration "
            f"(in {props_file.as_posix()})"
        )

    if clean:
        shutil.rmtree(output_dir, ignore_errors=True)

    lang_names = {
        book_lang: i18n.localize_lang() for book_lang, (_, i18n, _) in books.items()
    }

    for book_lang, (book, i18n, context) in books.items():
        png_textures = PNGTexture.get_lookup(context.textures)
        animations = list(AnimatedTexture.get_lookup(context.textures).values())

        render_book(
            props=props,
            pm=pm,
            lang=book_lang,
            lang_names=lang_names,
            book=book,
            i18n=i18n,
            templates=templates,
            output_dir=output_dir,
            site_path=site_path,
            all_metadata=all_metadata,
            png_textures=png_textures,
            animations=animations,
            allow_missing=allow_missing,
            version=plugin.mod_version,
        )

    logger.info("Done.")


@app.command()
def merge(
    *,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    release: ReleaseOption = False,
):
    # ensure at least the default language was built successfully
    if update_latest:
        assert_version_exists(root=src, version="latest")

    # TODO: figure out how to do this with pluggy (we don't have the props file here)
    # if is_release:
    #     assert_version_exists(src, GRADLE_VERSION)

    dst.mkdir(parents=True, exist_ok=True)

    # remove any stale data that we're about to replace
    delete_updated_books(src=src, dst=dst)
    if update_latest and release:
        delete_root_book(root=dst)

    # do the merge
    shutil.copytree(src=src, dst=dst, dirs_exist_ok=True)

    # rebuild the sitemap
    sitemap = load_sitemap(dst)
    dump_sitemap(dst, sitemap)


@app.command()
def serve(
    *,
    props_file: PropsOption,
    port: int = 8000,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    branch: BranchOption,
    release: bool = True,  # generally want --release for development, looks nicer
    clean: bool = False,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    book_root = dst
    relative_root = book_root.resolve().relative_to(Path.cwd())

    base_url = f"http://localhost:{port}"
    book_url = f"{base_url}/{relative_root.as_posix()}"

    repo_root = git_root(props_file.parent)
    asset_root = repo_root.relative_to(Path.cwd())

    os.environ |= {
        # prepend a slash to the path so it can find the texture in the local repo
        # eg. http://localhost:8000/_site/src/docs/Common/...
        # vs. http://localhost:8000/Common/...
        "DEBUG_GITHUBUSERCONTENT": f"{base_url}/{asset_root.as_posix()}",
        "GITHUB_PAGES_URL": book_url,
    }

    print("Exporting...")
    export(
        branch=branch,
        release=release,
        props_file=props_file,
        allow_missing=allow_missing,
        verbosity=verbosity,
    )

    print("Rendering...")
    render(
        branch=branch,
        props_file=props_file,
        output_dir=src,
        release=release,
        clean=clean,
        lang=lang,
        allow_missing=allow_missing,
        verbosity=verbosity,
    )

    print("Merging...")
    merge(
        src=src,
        dst=dst,
        release=release,
    )

    print(f"Serving web book at {book_url} (press ctrl+c to exit)\n")
    with HTTPServer(("", port), SimpleHTTPRequestHandler) as httpd:
        # ignore KeyboardInterrupt to stop Typer from printing "Aborted."
        # because it keeps printing after nodemon exits and breaking the output
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            pass


if __name__ == "__main__":
    app()
