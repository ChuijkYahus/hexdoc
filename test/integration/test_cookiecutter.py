# pyright: reportUnknownMemberType=false

import json
import subprocess
import sys
from pathlib import Path
from textwrap import dedent
from typing import Callable, Literal

from hexdoc.cli.app import render
from hexdoc.utils import JSONValue
from pytest import MonkeyPatch
from pytest_cookies.plugin import Cookies
from syrupy.assertion import SnapshotAssertion

from ..conftest import list_directory, nox_only

FileValue = JSONValue | tuple[Literal["a"], str] | Callable[[str | None], str]

FileTree = dict[str, "FileTree | FileValue"]


def write_file_tree(root: Path, tree: FileTree):
    for path, value in tree.items():
        path = root / path
        match value:
            case {**children} if path.suffix != ".json":
                # subtree
                path.mkdir(parents=True, exist_ok=True)
                write_file_tree(path, children)
            case {**json_data}:
                # JSON file
                path.parent.mkdir(parents=True, exist_ok=True)
                path.write_text(json.dumps(json_data, indent="  "))
            case tuple((mode, text)):
                # append to existing file
                with path.open(mode) as f:
                    f.write(dedent(text))
            case str() as text:
                # anything else - usually just text
                path.parent.mkdir(parents=True, exist_ok=True)
                path.write_text(dedent(text))
            case int() | float() | bool() | None:
                raise TypeError(
                    f"Type {type(value)} is only allowed in JSON data: {value}"
                )
            case fn:
                assert not isinstance(fn, (list, dict))
                path.parent.mkdir(parents=True, exist_ok=True)
                try:
                    current_value = path.read_text()
                except FileNotFoundError:
                    current_value = None
                path.write_text(dedent(fn(current_value)))


def run_pip(*args: str):
    return subprocess.run((sys.executable, "-m", "pip") + args, check=True)


@nox_only
def test_cookiecutter(
    cookies: Cookies,
    monkeypatch: MonkeyPatch,
    snapshot: SnapshotAssertion,
    path_snapshot: SnapshotAssertion,
    env_overrides: dict[str, str],
):
    result = cookies.bake(
        {
            "output_directory": "output",
            "modid": "mod",
            "pattern_regex": "hex_latest",
            "multiloader": False,
            "java_package": "com/package",
            "pattern_registry": "Patterns.java",
        }
    )

    assert result.exception is None
    assert result.project_path is not None

    monkeypatch.chdir(result.project_path)

    subprocess.run(["git", "init"], check=True)

    write_file_tree(
        result.project_path,
        {
            "src/generated/resources": {},
            "src/main/java/com/package/Patterns.java": "",
            "src/main/resources/assets/hexcasting/patchouli_books/thehexbook/en_us": {
                "categories/foo.json": {
                    "name": "hexdoc.mod.title",
                    "icon": "minecraft:amethyst_shard",
                    "description": "hexcasting.category.basics.desc",
                    "sortnum": 0,
                },
                "entries/bar.json": {
                    "name": "hexdoc.welcome.header",
                    "category": "hexcasting:foo",
                    "icon": "minecraft:textures/mob_effect/nausea.png",
                    "sortnum": 0,
                    "advancement": "hexcasting:y_u_no_cast_angy",
                    "pages": [
                        {
                            "type": "patchouli:text",
                            "text": "hexcasting.page.couldnt_cast.1",
                        },
                    ],
                },
            },
            ".env": "\n".join(
                f"{name}={value}" for name, value in env_overrides.items()
            ),
            "gradle.properties": (
                """\
                modVersion=1.0.0
                hexcastingVersion=0.11.1-7
                minecraftVersion=1.20.1
                """
            ),
            # TODO: remove when textures are fully implemented
            "doc/hexdoc.toml": (
                "a",
                """
                [textures]
                missing = [
                    "minecraft:*",
                    "hexcasting:*",
                ]
                """,
            ),
        },
    )

    output_dir = Path("_site")

    try:
        run_pip("install", ".", "--force-reinstall", "--no-deps")
        render(
            output_dir=output_dir,
            props_file=Path("doc/hexdoc.toml"),
            branch="main",
        )
    finally:
        run_pip("uninstall", "hexdoc-mod", "-y")

    path_snapshot._custom_index = "vlatestindex.html"  # pyright: ignore[reportPrivateUsage]
    assert output_dir / "v" / "latest" / "index.html" == path_snapshot

    assert list_directory(output_dir) == snapshot
