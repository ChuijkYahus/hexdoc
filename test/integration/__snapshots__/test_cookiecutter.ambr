# serializer version: 1
# name: test_cookiecutter
  list([
    'v',
    'v/latest',
    'v/latest/.sitemap-marker.json',
    'v/latest/apple-touch-icon-114x114.png',
    'v/latest/apple-touch-icon-120x120.png',
    'v/latest/apple-touch-icon-144x144.png',
    'v/latest/apple-touch-icon-152x152.png',
    'v/latest/apple-touch-icon-167x167.png',
    'v/latest/apple-touch-icon-180x180.png',
    'v/latest/apple-touch-icon-57x57.png',
    'v/latest/apple-touch-icon-60x60.png',
    'v/latest/apple-touch-icon-72x72.png',
    'v/latest/apple-touch-icon-76x76.png',
    'v/latest/favicon-16x16.png',
    'v/latest/favicon-180x180.png',
    'v/latest/favicon-196x196.png',
    'v/latest/favicon-32x32.png',
    'v/latest/favicon-64x64.png',
    'v/latest/favicon-96x96.png',
    'v/latest/favicon.ico',
    'v/latest/icon.png',
    'v/latest/index.css',
    'v/latest/index.html',
    'v/latest/index.js',
    'v/latest/mstile-270x270.png',
    'v/latest/mstile-310x150.png',
    'v/latest/mstile-310x310.png',
    'v/latest/mstile-70x70.png',
    'v/latest/textures.css',
  ])
# ---
